[#17, #18] 사용자 회원가입, 로그인 기능 구현

tinajeong opened the pull request:
### 🍀 목적
- 사용자 회원가입, 로그인 기능 구현

### 🌹 추가사항(있다면 적고 없다면 적지 않는다.)
- 회원가입, 로그인 기능 추가

### ☕ 변경사항(있다면 적고 없다면 적지 않는다.)
- `User` 클래스 변경

### ❓ pr 포인트
- 테스트 어떤가요.


### 📢 Help
- 다음 스프린트에서 삭제되거나 수정될 코드: `UserFindResponse` DTO , `Role` 클래스, `HttpSession` 관련 로직 

- logout 실패 시 403(Forbidden)을 사용해봤습니다!
  - 유효하지 않은 로그아웃 액션에 대해 보통은 `401`(Unauthorized)로 처리합니다.
  -  **서버가 로그아웃 요청을 이해하지만 금지되었다**는 의미를 명확히 전달하는 코드에 대해 좀더 고민해봤습니다.
  - 아래 시나리오에 대해 서버가 명확히 파악하고 있다고 생각했습니다.
  - 이번 스프린트에서는 `HttpSession`을 이용해 로그아웃을 처리합니다. 이 맥락에서는 발생할 수 있는 시나리오의 개수가 다음과 같이 한정되어 있다고 생각했습니다.
```markdown
    1. 세션 타임아웃이 일어난 경우
    2. 이미 무효화된 세션에 대해 로그아웃 요청을 한 경우
    3. 존재한 적 없는 세션에 대해 로그아웃 요청을 한 경우  
```
#resolves #17 , #18<!-- pr이 머지되면 이슈가 자동으로 close되게 합니다. -->

tinajeong committed:
chore : User 변경 사항 반영
- 생일 LocalDate로 변경
- 닉네임 최소 글자 수 2개로 변경
- 비밀번호 검증 위해 change method 추가

tinajeong committed:
feat : 회원가입, 로그인, 유저 조회 기능 구현

tinajeong committed:
docs : API 문서 경로 추가, user http 파일 추가

tinajeong committed:
chore : API 설명 추가

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+		this.userMapper = userMapper;
+	}
+
+	@PostMapping(value = {"/auth/signup"})
+	public ResponseEntity<String> signup(@RequestBody UserCreateRequest userCreateRequest) {
+		this.userServiceDefault.register(userCreateRequest);
+		return new ResponseEntity<>("Sign Up Completed", HttpStatus.CREATED);
+	}
+
+	@GetMapping(value = {"users/{loginId}"})
```
auth로 path를 나눠 구현할거면 AuthController라고 따로 Controller를 하나 만드셔도 될 것 같습니다. 
Controller가 여러 path에 대해 책임이 많아져 뚱뚱해지는것 같아요

youngjijang replied:
user path 앞에 슬래시가 빠졌네요.

/users/{loginId}
/users/me

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+			.orElseGet(() -> ResponseEntity.status(HttpStatus.NOT_FOUND).build());
+	}
+
+	@PostMapping(value = {"/auth/login"})
+	public ResponseEntity<String> login(@RequestBody UserLoginRequest request, HttpSession session) {
+		Optional<User> user = this.userServiceDefault.login(request.getLoginId(), request.getPassword());
+		if (user.isPresent()) {
+			session.setAttribute("user", user.get());
+			return new ResponseEntity<>("Login Succeeded", HttpStatus.OK);
+		}
+		return new ResponseEntity<>("Login Failed", HttpStatus.UNAUTHORIZED);
+	}
+
+	@PostMapping(value = {"/auth/logout"})
+	public ResponseEntity<Void> logout(HttpSession session) {
+		User user = (User)session.getAttribute("user");
+		if (user != null) {
+			session.removeAttribute("user");
+			session.invalidate();
+			return ResponseEntity.noContent().build();
+		} else {
+			return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
+		}
+
+	}
```
위와 동일합니다!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+	private final UserServiceDefault userServiceDefault;
+	private final UserMapper userMapper;
+
+	public UserController(UserServiceDefault userServiceDefault, UserMapper userMapper) {
+		this.userServiceDefault = userServiceDefault;
+		this.userMapper = userMapper;
+	}
+
```
생성자에 따로 비즈니스 로직이 없다면 `@RequiredArgumentConstructor` 를 사용해도 무방할 것같아요!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+		this.userMapper = userMapper;
+	}
+
+	@PostMapping(value = {"/auth/signup"})
+	public ResponseEntity<String> signup(@RequestBody UserCreateRequest userCreateRequest) {
+		this.userServiceDefault.register(userCreateRequest);
+		return new ResponseEntity<>("Sign Up Completed", HttpStatus.CREATED);
+	}
```
따로 RedirectURI 등을 주지 않고 메시지로 돌려준 이유가 궁금합니다.

tinajeong replied:
현재로서는 Redirection이 불필요하다고 생각했습니다!

youngjijang commented on http/user.http:
```diff
@@ -0,0 +1,52 @@
+### 회원가입 성공
```
👍🏻

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+
+@RequestMapping("/api/v1")
+@RestController
+@Slf4j
+public class UserController {
+	private final UserServiceDefault userServiceDefault;
```
전체적으로 메서드에 `@Valid` 어노테이션이 다 빠진것 같아요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	@PostMapping(value = {"/auth/login"})
+	public ResponseEntity<String> login(@RequestBody UserLoginRequest request, HttpSession session) {
```
Entity Class는 Transaction 밖으로 꺼내지 않기로 하였고, LazyLording 등의 이슈를 피하려면 Service 로직 안으로 집어넣는게 좋아보입니다!

boolean 반환형인 isPresent()로 if문을 분기탈 경우,  Optional<엔티티> 보다는 isSuccess같은 것을 사용하면 어떨까요?

tinajeong committed:
Merge branch 'develop' into feature/user

# Conflicts:
#	src/docs/asciidoc/index.adoc

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	@GetMapping(value = {"users/me"})
+	public ResponseEntity<UserFindResponse> getMe(HttpSession session) {
+		User user = (User)session.getAttribute("user");
+		if (user != null) {
+			return ResponseEntity.ok(userMapper.entityToFindResponse(user));
+		} else {
+			return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+		}
+	}
```
여기도 마찬가지로 Entity Class를 직접적으로 Transaction 밖으로 노출시켜서 접근하는것보다는, 
Session이나 Security를 사용해야 할 경우 Adaptor 패턴으로 인증 객체를 만드는게 어떨끼요?

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/dto/UserCreateRequest.java:
```diff
@@ -21,16 +27,18 @@ public class UserCreateRequest {
 
 	@NotNull(message = "이메일은 필수 입력사항입니다.")
 	@Email(message = "이메일 형식을 확인해주세요.")
 	private String email;
+
+	private Role role;
```
Role은 Null을 허용하나요 ??

tinajeong replied:
Role은 삼항연산자를 이용해 default값이 Role.USER 였으나 MapStruct를 적용하면서 사라졌습니다. 
Role enum 자체가 다음 스프린트에서 대체될 임시 장치라 null을 허용한 상태로 두었습니다~

youngjijang commented on src/main/java/com/prgrms/bdbks/domain/user/repository/UserRepository.java:
```diff
@@ -0,0 +1,18 @@
+@Repository
+public interface UserRepository
+	extends JpaRepository<User, String> {
+	Optional<User> findByLoginId(String var1);
```
매개변수명을 var로 지정하신 이유가 있나요??

youngjijang commented on src/main/java/com/prgrms/bdbks/domain/user/service/UserServiceDefault.java:
```diff
@@ -0,0 +1,49 @@
+		this.userRepository = userRepository;
+		this.passwordEncoder = passwordEncoder;
+		this.userMapper = userMapper;
+	}
```
@RequiredArgsConstructor 를 사용하면 편하게 주입이 가능할 것 같아요!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/repository/UserRepository.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository
+	extends JpaRepository<User, String> {
```
줄바꿈 위로가면 좋을것같아요!

tinajeong replied:
decompile해 사용하는 과정에서 더블체크를 하지 않았네요! 수정하겠습니다.

rlarltj replied:
@tinajeong 여기 String이 아니라 Long타입이어야 하지 않을까요 !??

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/repository/UserRepository.java:
```diff
@@ -0,0 +1,18 @@
+@Repository
+public interface UserRepository
+	extends JpaRepository<User, String> {
+	Optional<User> findByLoginId(String var1);
+
+	@Query(value = "SELECT u FROM User u WHERE u.loginId = ?1 AND u.password = ?2")
+	Optional<User> findByLoginIdAndPassword(String var1, String var2);
+}
```
변수명이 수정이 필요해요. 

`@Param` 이용해서 이름 방식으로 직접 매핑하는게 좋을것같아요

tinajeong replied:
위와 동일합니다!

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/service/UserServiceDefault.java:
```diff
@@ -0,0 +1,49 @@
+import org.springframework.stereotype.Service;
+
+@Service
+public class UserServiceDefault implements UserService {
+	private final UserRepository userRepository;
+	private final PasswordEncoder passwordEncoder;
+	private final UserMapper userMapper;
+
```
Default는 prefix로 가는게 좋을것같아요!

tinajeong replied:
팀 기술문서에 적어두었습니다~!

>  Service를 상속 받는 기본 서비스 구현체는 Default를 prefix로 둔다.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/service/UserServiceDefault.java:
```diff
@@ -0,0 +1,49 @@
+		this.passwordEncoder = passwordEncoder;
+		this.userMapper = userMapper;
+	}
+
+	@Override
+	public User register(UserCreateRequest userCreateRequest) {
+		User user = userMapper.createRequestToEntity(userCreateRequest);
+		user.changePassword(this.passwordEncoder.encode(user.getPassword()));
+		return this.userRepository.save(user);
+	}
+
+	@Override
+	public Optional<User> findUser(String loginId) {
+		return this.userRepository.findByLoginId(loginId);
+	}
+
+	@Override
+	public Optional<User> login(String loginId, String password) {
+		Optional<User> user = this.findUser(loginId);
+		if (user.isPresent()) {
+			String encodedPassword = user.get().getPassword();
+			if (this.passwordEncoder.matches(password, encodedPassword)) {
+				return this.userRepository.findByLoginIdAndPassword(loginId, encodedPassword);
+			}
+			return Optional.empty();
+		}
+		return Optional.empty();
+	}
+}
```
`@Transactional` 이 없는것같아요.

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/service/UserServiceDefault.java:
```diff
@@ -0,0 +1,49 @@
+		return this.userRepository.findByLoginId(loginId);
+	}
+
+	@Override
+	public Optional<User> login(String loginId, String password) {
+		Optional<User> user = this.findUser(loginId);
+		if (user.isPresent()) {
+			String encodedPassword = user.get().getPassword();
+			if (this.passwordEncoder.matches(password, encodedPassword)) {
+				return this.userRepository.findByLoginIdAndPassword(loginId, encodedPassword);
+			}
+			return Optional.empty();
+		}
+		return Optional.empty();
+	}
```
현재 Session에 User를 저장하기 때문에 컨트롤러로 직접 User Entity를 리턴하는것 같아요.
이러면 우리가 이야기 했던 LazyLoading등 여러 문제가 있을수가 있어서 방법을 여러개 생각해볼 수 있을것 같아요

1. DTO 사용
2. UserEntity를 담은 Adapter패턴을 이용하여 Authentication용 객체를 만들어 반환하여 Session에 저장 
3. id만 반환 후 세션에 추가 로직 적용하기

devYSK commented on src/test/java/com/prgrms/bdbks/domain/user/entity/controller/UserControllerMvcTest.java:
```diff
@@ -0,0 +1,155 @@
+
+@AutoConfigureRestDocs
+@WebMvcTest(controllers = UserController.class)
+@AutoConfigureMockMvc(addFilters = false)
```
제가 401,403 에러를 해결하기 위한 방법으로 addFilters = false와 .with(csrf())를 두 방법을 기술문서에  작성하였는데,
`addFilters = false`는 사용하지 않는것이 좋을것 같아요. 
Filter들을 다 사용 안하게 되버리니까요
with(csrf()를 사용하는것이 어떨까요!

tinajeong replied:
` .with(csrf())`를 사용하면 토큰값이 넘어가네요! 좋은 추천감사합니다!

devYSK commented on src/test/java/com/prgrms/bdbks/domain/user/entity/controller/UserControllerMvcTest.java:
```diff
@@ -0,0 +1,155 @@
+		emptyUser = new EmptyUser();
+	}
+
+	@DisplayName("등록 - 사용자 가입에 성공하고 201 코드를 리턴한다.")
+	@Test
+	void signup_success() throws Exception {
+		String USER_NAME = "blackdog";
+		String USER_PHONE = "01012341234";
+		String USER_EMAIL = "blackdog@blackdog.com";
+		UserCreateRequest userCreateRequest = new UserCreateRequest(USER_LOGIN_ID, USER_PASSWORD, USER_NAME,
+			USER_BIRTH_DATE, USER_PHONE, USER_EMAIL, USER_ROLE);
```
가입에 실패하는 경우는 어떤 경우가 있으면 좋을까요?

저는 Validation Fail, 중복 이메일 등을 떠올려볼 수 있을것같아요

tinajeong replied:
Validation은 통합테스트 추가해서 검증하고 중복 이메일 케이스는 추가해볼게요~

devYSK reviewed (changes requested):
찬미님 고생 많으셨습니다!

수정할 부분이 필요해 보여 리뷰 몇자 남겼습니다 확인 부탁드려요. 감사합니다

tinajeong committed:
refact :
- @RequiredArgsConstructor 사용
- DefaultUserService renaming
- test 시 with csrf 사용
- UserService에 @Transactional 사용
- Controller에 entity 노출 자제

tinajeong committed:
Merge branch 'develop' into feature/user

tinajeong committed:
refact : 아이디 중복 검증 로직 추가

tinajeong committed:
Merge remote-tracking branch 'origin/feature/user' into feature/user

devYSK commented on src/test/java/com/prgrms/bdbks/domain/user/entity/UserTest.java:
```diff
@@ -94,7 +94,7 @@ void validateNickname_EmptyValue_ExceptionThrown(String invalidPassword) {
 	}
 
```
LGTM

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/service/DefaultUserService.java:
```diff
@@ -0,0 +1,57 @@
+	public Optional<User> login(String loginId, String password) {
+		Optional<User> user = this.userRepository.findByLoginId(loginId);
+		if (user.isPresent()) {
+			String encodedPassword = user.get().getPassword();
+			if (this.passwordEncoder.matches(password, encodedPassword)) {
+				return this.userRepository.findByLoginIdAndPassword(loginId, encodedPassword).map(UserAdapter::new);
```
패스워드가 유저의 패스워드랑 일치하는지는 유저 도메인의 책임으로 볼 수 있을것 같아요.

패스워드 인코더를 유저로 넘겨 유저 도메인 내에서 확인하는건 어떨까요? 
추가적으로, 패스워드 인코더는 인터페이스이니까 그러면 다른 구현체로 바뀌어도 문제는 없겠죠?

tinajeong replied:
시큐리티 적용하면서 함께 수정하겠습니다~

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/service/DefaultUserService.java:
```diff
@@ -0,0 +1,57 @@
+	@Transactional
+	public User register(UserCreateRequest userCreateRequest) {
+		User user = userMapper.createRequestToEntity(userCreateRequest);
+		user.changePassword(this.passwordEncoder.encode(user.getPassword()));
+		if (!this.userRepository.existsByLoginId(user.getLoginId())) {
+			return new UserAdapter(this.userRepository.save(user));
```
마찬가지로 여기도 패스워드를 암호화 하는 책임은 유저한테 있을 수 있을것 같아요. 
도메인안에 로직을모아둠으로써 응집도가 향상되고, 패스워드는 유저의 것이고 유저한테 책임을 모아둘 수 있겠죠? 

패스워드 인코더를 유저로 넘겨 유저 도메인 내에서 암호화 하는건 어떨까요? 
추가적으로, 패스워드 인코더는 인터페이스이니까 그러면 다른 구현체로 바뀌어도 문제는 없겠죠? 
패스워드가 일치하는지도 같이 볼 수 있을것 같네요

devYSK commented on src/main/java/com/prgrms/bdbks/domain/user/entity/UserAdapter.java:
```diff
@@ -0,0 +1,38 @@
+@NoArgsConstructor
+public class UserAdapter extends User {
+
+	String loginId;
+
+	String nickname;
+
+	String password;
+	String email;
+
+	Role role;
+
+	public UserAdapter(User user) {
```
접근제한자가 빠져있네요!

tinajeong replied:
위와 같습니다~

devYSK reviewed (changes requested):
고생많으셨습니다! 중요부분 몇군데 수정이 필요할것같아용.

gw0215 reviewed (approved):
고생하셨습니다.

tinajeong commented:
머지하겠습니다.
